### PR TITLE
Add simple OpenTelemetry tracer

### PR DESCRIPTION
**Pull request by yujonglee**
## Title
Add simple OpenTelemetry tracer

## Type
🆕 New Feature

## Changes
Add OpenTelemetry logger

Reference:
https://opentelemetry.io/docs/languages/python/instrumentation/

### CODE REVIEW

File: litellm/integrations/opentelemetry.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+from opentelemetry.sdk.trace.export import (
+    BatchSpanProcessor,
+    ConsoleSpanExporter,
+)
+
+
+LITELLM_TRACER_NAME = "litellm"
+LITELLM_RESOURCE = {"service.name": "litellm"}
+
+MOCK_TRACE_PARENT = {"traceparent": "SOMETHING_FROM_PROXY_REQUEST"}
```
Comment (yujonglee) [inline litellm/integrations/opentelemetry.py]: Assuming that we will use otel for proxy, how to get `proxy_server_request.header`?

File: litellm/integrations/opentelemetry.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+MOCK_SPAN_NAME = "TODO"
```
Comment (yujonglee) [inline litellm/integrations/opentelemetry.py]: What is good way to name span?
Is there some litellm's internal state that represent "completion/chat/embedding..."?

File: pyproject.toml [hunk 3]
```diff
@@ -40,6 +40,9 @@ azure-identity = {version = "^1.15.0", optional = true}
 azure-keyvault-secrets = {version = "^4.8.0", optional = true}
 google-cloud-kms = {version = "^2.21.3", optional = true}
 resend = {version = "^0.8.0", optional = true}
+opentelemetry-api = "^1.25.0"
+opentelemetry-sdk = "^1.25.0"
+opentelemetry-exporter-otlp = "^1.25.0"
```
Comment (yujonglee) [inline pyproject.toml]: Otel should not included in lib, but included in proxy docker image, right?

File: litellm/integrations/opentelemetry.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        )
+        span.set_status(Status(StatusCode.ERROR))
+        self._set_attributes(span, kwargs, response_obj)
+        span.end(end_time=self._to_ns(end_time))
+
+    def _to_ns(self, dt):
+        return int(dt.timestamp() * 1e9)
+
+    def _set_attributes(self, span, kwargs, response_obj):
+        keys = ["model", "api_base", "api_version"]
```
Comment (yujonglee) [inline litellm/integrations/opentelemetry.py]: What will be good other attributes?

File: pyproject.toml (hunk 3 above)
Comment (ghost) [reply pyproject.toml]: yep - add this in the requirements.txt

File: litellm/integrations/opentelemetry.py (hunk 4 above)
Comment (ghost) [reply litellm/integrations/opentelemetry.py]: what does attributes mean here exactly? is the prometheus logger a good reference? @yujonglee 

For LLM API Calls - https://github.com/BerriAI/litellm/blob/c280de3f4559faac2d76e7909273711717e3a3f0/litellm/integrations/prometheus.py#L13

For system health metrics - https://github.com/BerriAI/litellm/blob/c280de3f4559faac2d76e7909273711717e3a3f0/litellm/integrations/prometheus_services.py#L15

File: litellm/integrations/opentelemetry.py (hunk 2 above)
Comment (ghost) [reply litellm/integrations/opentelemetry.py]: does each span name need to be unique? 

langfuse integration for naming traces might be relevant - .https://github.com/BerriAI/litellm/blob/c280de3f4559faac2d76e7909273711717e3a3f0/litellm/integrations/langfuse.py#L334

the logging object has a value `call_type` which we use

File: litellm/integrations/opentelemetry.py (hunk 1 above)
Comment (ghost) [reply litellm/integrations/opentelemetry.py]: it's available in the call metadata https://github.com/BerriAI/litellm/blob/c280de3f4559faac2d76e7909273711717e3a3f0/litellm/integrations/langfuse.py#L428

File: litellm/integrations/opentelemetry.py (hunk 4 above)
Comment (yujonglee) [reply litellm/integrations/opentelemetry.py]: @krrishdholakia 

https://opentelemetry.io/docs/concepts/signals/traces/#attributes

I think `["model", "api_base", "api_version"]` is enough for now.

File: litellm/tests/test_proxy_server.py [hunk 5]
```diff
@@ -608,3 +602,28 @@ def test_load_router_config(mock_cache, fake_env_vars):
...
 
 # test_load_router_config()
+
+from litellm.integrations.opentelemetry import OpenTelemetry, OpenTelemetryConfig
+
+
+@mock_patch_acompletion()
+def test_otel_with_proxy_server(mock_acompletion, client_no_auth, capsys):
+    litellm.callbacks = [OpenTelemetry(OpenTelemetryConfig(exporter="console"))]
+
```
Comment (yujonglee) [inline litellm/tests/test_proxy_server.py]: It seems like this callback is not even run when invoking this test.
what is correct way to test callback with proxy?

**Review by yujonglee — COMMENTED** [review]
@krrishdholakia 

I think the `opentelemetry.py` itself is complete.

Few missing things - 

- proxy test with otel callback
- use this callback when we have `otel=true` in config (https://litellm.vercel.app/docs/simple_proxy_old_doc#set-opentelemetry-oteltrue-on-the-proxy-configyaml)
- update docs

**Review by ishaan-jaff — COMMENTED** [review]
question - how do I use this to export to an OTEL collector ?

**Comment by yujonglee** [comment]
@ishaan-jaff 

1. Test if it works for console
```bash
# env
OTEL_EXPORTER="console"
```

You will see something like this:
```json
{
    "name": "TODO",
    "context": {
        "trace_id": "0x5f0bc3696e35846b1eb84ffc8a90f158",
        "span_id": "0x4e2c51583de4c9e3",
        "trace_state": "[]"
    },
    "kind": "SpanKind.INTERNAL",
    "parent_id": null,
    "start_time": "2024-06-02T04:35:25.609743Z",
    "end_time": "2024-06-02T04:35:26.205273Z",
    "status": {
        "status_code": "OK"
    },
    "attributes": {
        "model": "gpt-3.5-turbo"
    },
    "events": [],
    "links": [],
    "resource": {
        "attributes": {
            "service.name": "litellm"
        },
        "schema_url": ""
    }
}
```

2. For sending collector
```bash
# env
OTEL_EXPORTER="otlp_http"
OTEL_ENDPOINT="<collector_endpoint>"
OTEL_BEARER_TOKEN="<collector_auth>"
```

---
Note that in litellm docs(https://docs.litellm.ai/docs/simple_proxy_old_doc#set-destination-paths-on-opentelemetry-collecter), it only use `otlp_grpc`.

For accepting both `http` and `grpc`, we need something like this:

```yaml
receivers:
  otlp:
    protocols:
      http:
        endpoint: 0.0.0.0:4318
      grpc:
        endpoint: 0.0.0.0:4317
```

**Comment by ishaan-jaff** [comment]
@yujonglee what's pending to merge this in ?

**Comment by yujonglee** [comment]
Screenshot 2024-06-04 at 1 56 45 PM

**Review by ishaan-jaff — APPROVED** [review]

**Comment by ishaan-jaff** [comment]
main is stable - merging this

**Pull request merged**